specification.md

```diff
@@ -3880,6 +3903,286 @@ execution of the state, runtimes should not create a new instance of this workfl
+    nextState: SomeNextWorkflowState
+- name: CancelPurchase
+  type: operation
+  usedForCompensation: true
+  actions:
+  - functionRef:
+      refName: CreditCustomerFunction
+      parameters:
+        customerid: "{{ $.purchase.customerid }}"
+        amount: "{{ $.purchase.amount }}"
+  - functionRef:
+      refName: SendPurchaseCancellationEmailFunction
+      parameters:
+        customerid: "{{ $.purchase.customerid }}"
+```
+</td>
+</tr>
+</table>
+
+In this example our "NewItemPurchase" [event state](#Event-state) waits for a "NewPurchase" event and then 
+debits the customer and sends them a purchase confirmation email. It defines that it's compensated by the 
+"CancelPurchase" [operation state](#Operation-state) which performs two actions, namely credits back the 
+purchase amount to customer and sends them a purchase cancellation email.
+
+#### Triggering Compensation
+
+As previously mentioned, compensation must be explicitly triggered by the workflows control-flow logic. 
+This can be done via [transition](#Transition-definition) and [end](#End-definition) definitions.
+
+Let's take a look at each:
+
+1. Compensation triggered on transition:
+
+<table>
+<tr>
+    <th>JSON</th>
+    <th>YAML</th>
+</tr>
+<tr>
+<td valign="top">
+
+```json
+{
+  "transition": {
+      "compensateBefore": true,
+      "nextState": "NextWorkflowState"
+  }
+}
+```
+</td>
+<td valign="top">
+
+```yaml
+transition:
+  compensateBefore: true
+  nextState: NextWorkflowState
+```
+</td>
+</tr>
+</table>
+
+Transitions can trigger compensations by specifying the `compensateBefore` property and setting it to `true`.
+This means that before the transition is executed (workflow continues its execution to the "NextWorkflowState" in this example),
+workflow compensation must be performed. 
+
+2. Compensation triggered by end definition:
+
+<table>
+<tr>
+    <th>JSON</th>
+    <th>YAML</th>
+</tr>
+<tr>
+<td valign="top">
+
+```json
+{
+  "end": {
+    "compensateBefore": true,
+    "kind":"default"
+  }
+}
+```
+</td>
+<td valign="top">
+
+```yaml
+end:
+  compensateBefore: true
+  kind: default
+```
+</td>
+</tr>
+</table>
+
+End definitions can trigger compensations by specifying the `compensateBefore` property and setting it to `true`.
+This means that before workflow finishes its execution workflow compensation must be performed. Note that 
+in case when the end definition has its "kind" set to "event", compensation must be performed before 
+producing the specified events and ending workflow execution.
+
+#### Compensation Execution Details
+
+Now that we have seen how to define and trigger compensation, we need to go into details on how compensation should be executed.
+Compensation is performed on all already successfully completed states (that define `compensatedBy`) in **reverse** order.
+Compensation is always done in sequential order, and should not be executed in parallel.
+
+Let's take a look at the following workflow image:
+
+<p align="center">
```

Comment:
Nit; when we provide an image, should we also be in the habit of providing some sample json/yaml definition?  Perhaps hidden behind an 'expand definition' link so it's not too cluttered unless the reader wants to read/copy the definition?

Reply:
Good point. I will add asap.